pkg/kubelet/container/helpers.go

```diff
@@ -117,6 +117,22 @@ func HashContainer(container *v1.Container) uint64 {
 	return uint64(hash.Sum32())
 }
 
+func HashContainerForContainerChange(container *v1.Container) uint64 {
+	hash := fnv.New32a()
+	containerJSON, _ := json.Marshal(pickFieldToHash(container))
+	hashutil.DeepHashObject(hash, containerJSON)
+	return uint64(hash.Sum32())
+}
+
+// pickFieldToHash pick fields that will affect the running status of the container for hash,
+// currently this field range only contains `image`.
+func pickFieldToHash(container *v1.Container) *v1.Container {
```

Comment:
structurally, this looks plausible to me

I'd like node folks to weigh in on whether there are any other fields that should be conditionally included, especially if the VPA feature is enabled

Reply:
@kubernetes/sig-node-pr-reviews Would appreciate any advice.

Reply:
There is [this](https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/pkg/apis/core/validation/validation.go#L5250-L5259) validation logic which correlates closely to [Pod update and replacement](https://kubernetes.io/docs/concepts/workloads/pods/#pod-update-and-replacement) documentation. @liggitt map suggestion needs to take into account the fields found in the validation logic+documentation.

Reply:
Hmm. Actually, after thinking about this a bit more, we probably want to leave these additional fields out.

I do think we need to mixin the UID within the hash. Thoughts?

Reply:
only image and resources (when the VPA feature is enabled) can mutate inside containers:

https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/pkg/apis/core/validation/validation.go#L5147-L5179

my understanding is that `resources` changing is not expected to instantly restart the container (as would happen if the hash changed) but restart in a controlled way via a separate code path

Reply:
> I do think we need to mixin the UID within the hash. Thoughts?

uid wasn't included before... this is a hash of just the `container` substanza

Reply:
> My general concern is that this change might effect static pods.

Hmm... static pod ingestion generates uids for static pods based on a full hash of the full pod content, which should make modified static manifests show up as a completely different pod uid, right?:
https://github.com/kubernetes/kubernetes/blob/be4b7176dc131ea842cab6882cd4a06dbfeed12a/pkg/kubelet/config/common.go#L58-L71

Also, the kubelet isn't really expecting to handle changes to a container other than image on an update, so I think if static pods are making changes the API wouldn't permit AND hard-coding the uid to skip the unique uid generation for static pods, they're in bug territory already